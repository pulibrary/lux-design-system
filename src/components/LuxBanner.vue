<template>
  <div v-if="show" :class="['lux-banner', { 'lux-banner-fullscreen': isFullScreen }]">
    <div class="lux-announcement open" role="alert">
      <lux-wrapper maxWidth="1125">
        <button
          v-if="dismissible"
          type="button"
          class="lux-announcement__close"
          data-dismiss="alert"
          aria-label="Close"
          @click="hideBanner()"
        >
          <span aria-hidden="true">×</span>
        </button>
        <!-- @slot The content for your banner. -->
        <slot>{{ bannerMessage }}</slot>
      </lux-wrapper>
    </div>
  </div>
</template>

<script>
// import VueCookies from "vue-cookies"
import { useCookies } from "vue3-cookies"

/**
 * Banners are used to provide timely, general information to users of a website or app.
 */
export default {
  name: "LuxBanner",
  status: "prototype",
  release: "2.16.3",
  type: "Element",
  metaInfo: {
    title: "Global alert message to user.",
    htmlAttrs: {
      lang: "en",
    },
  },
  setup() {
    const { cookies } = useCookies()
    return { cookies }
  },
  data: function () {
    return {
      bannerMessage: "This is the announcement description.",
      show: true,
    }
  },
  computed: {
    isFullScreen() {
      return this.fullscreen === "true" ? true : false
    },
  },
  props: {
    /**
     * Allows the banner to act as an overlay, near the top of the screen.
     */
    fullscreen: {
      type: Boolean,
      default: false,
    },
    /**
     * User can manually hide the notification.
     */
    dismissible: {
      type: Boolean,
      default: false,
    },
  },
  methods: {
    hideBanner() {
      this.show = false
      this.cookies.set("showBanner", false)
    },
  },
  created: function () {
    if (this.cookies.isKey("showBanner")) {
      this.show = this.cookies.get("showBanner") === "true"
    }
  },
}
</script>

<style lang="scss">
@import "../assets/styles/mixins.scss";
@import "../assets/styles/variables.css";

.lux-banner-fullscreen {
  position: fixed;
  top: 5vh;
  left: 50%; /* move the left edge to the center … */
  margin-left: -25vw !important; /* … and move it to the left half the box’ width. */
  z-index: 9999;
  width: 50vw;
}

.lux-banner {
  @include reset;
  font-family: var(--font-family-text);
}

.lux-banner-dismissible > span {
  width: calc(100% - 25px);
  display: inline-block;
}

.lux-banner-dismissible .lux-banner__close {
  position: absolute;
  top: -4px;
  right: -5px;
  padding: 0.75rem 1.25rem;
  color: inherit;
}

.lux-announcement {
  position: relative;
  background: #1da3ff;
  background: var(--color-bleu-de-france-dark);
  color: white;
  padding: 1.5rem 0.5rem 0.5rem 0.5rem;
  transition: 0.2s;
  z-index: 1;

  a {
    color: white;
    transition: 0.25s ease;

    &:hover {
      opacity: 0.75;
    }
  }

  button.lux-announcement__close {
    padding: 0;
    cursor: pointer;
    background: 0 0;
    border: 0;
    -webkit-appearance: none;
  }

  &__close {
    position: absolute;
    right: 15px;
    top: 5px;
    font-size: 18px;
    width: 25px;
    height: 25px;
    font-weight: 600;
    text-align: center;
    cursor: pointer;
    color: white;
    opacity: 0.65;
    transition: 0.25s ease;

    &:hover {
      opacity: 1;
    }
  }
}

.open {
  top: 0;
}
</style>

<docs>
  ```jsx
  <div>
    <lux-banner dismissible>
      <h2>Fall Semester Announcements</h2>
      <p>We’re currently in beta for students and will be introducing faculty, advisor and staff functionality in the coming months.</p>
    </lux-banner>
  </div>
  ```
</docs>
