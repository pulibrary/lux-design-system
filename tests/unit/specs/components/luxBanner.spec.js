import { mount } from "@vue/test-utils"
import { nextTick } from "vue"
import LuxBanner from "@/components/LuxBanner.vue"

describe("LuxBanner.vue", () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(LuxBanner, {
      slots: {
        default: "<p>Here's some info for you.</p>",
      },
      stubs: ["wrapper"],
    })
  })

  it("should have an alert message", () => {
    const el = wrapper.find(".lux-banner")
    expect(el.text()).toBe("Here's some info for you.")
  })

  it("should have a button when dismissible", async () => {
    wrapper.setProps({ dismissible: true })
    await nextTick()
    const button = wrapper.find("button")
    expect(wrapper.vm.dismissible).toBe(true)
    expect(button.exists()).toBe(true)
  })

  it("should be dismissible on click", async () => {
    wrapper.setProps({ dismissible: true })
    const slot_text = wrapper.find("p")
    expect(slot_text.exists()).toBe(true)
    await nextTick()
    const button = wrapper.find("button")
    button.trigger("click")
    await nextTick()
    const p = wrapper.find("p")
    expect(p.exists()).toBe(false)
  })

  it("should hide banner when hideBanner() is called", () => {
    wrapper.vm.show = true
    wrapper.vm.hideBanner()
    expect(wrapper.vm.show).toBe(false)
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
