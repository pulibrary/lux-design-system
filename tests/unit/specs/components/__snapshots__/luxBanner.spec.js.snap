// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuxBanner.vue has the expected html structure 1`] = `<!--v-if-->`;
